README.md
# @atomist/npm-automations

[![Build Status](https://travis-ci.org/atomist/npm-automations.svg?branch=master)](https://travis-ci.org/atomist/npm-automations)

[Atomist][atomist] automations for interactive with NPM modules using
the [`@atomist/automation-client`][client] node module to implement a
local client that connects to the Atomist API.

[client]: https://github.com/atomist/automation-client-ts (@atomist/automation-client Node Module)

## Prerequisites

### Access to Atomist testing environment

To get access to this preview, please reach out to members of Atomist
in the `#support` channel of [atomist-community Slack team][slack].

You'll receive an invitation to a Slack team and GitHub organization
that can be used to explore this new approach to writing and running
automations.

### Node.js

You will need to have [Node.js][node] installed.  To verify that the
right versions are installed, please run:

```
$ node -v
v8.4.0
$ npm -v
5.4.1
```

[node]: https://nodejs.org/ (Node.js)

### Cloning the repository and installing dependencies

To get started run the following commands:

```
$ git clone git@github.com:atomist/npm-automations.git
$ cd npm-automations
$ npm install
```

### Configuring your environment

For the client to connect and authenticate to the Atomist API, a
GitHub personal access token is required. Additionally the API
is only allowing members of a GitHub team called `atommist-automation`
to successfully authenticate and register a new client.

Please create a team in your GitHub organization with the name
`atomist-automation` and add the user you want to use to the team.

After that you can create a personal access token with `read:org`
scope at https://github.com/settings/tokens.

Once you obtained the token, make it available to the client by
exporting it into a environment variable:

```
$ export GITHUB_TOKEN=<your token goes here>
```

Alternatively you can also place the token in `src/atomist.config.ts`.

## Starting up the automation-client

To start the client, run the following command:

```
$ npm run start
```

## Invoking a command handler from Slack

This project contains the code to create and respond to a simple
`hello world` bot command.  The code that defines the bot command and
implements responding to the command, i.e., the _command handler_, can
be found in [`HelloWorld.ts`][hello].  Once you have your local
automation client running (the previous step in this guide), you can
invoke the command handler by sending the Atomist bot the command in
the `#general` channel of the testing Slack team:

```
@atomist hello world
```

Once you've submitted the command in Slack, you'll see the incoming
and outgoing messages show up in the logs of your locally running
automation-client.  Ultimately, you should see the response from the
bot in Slack.

[hello]: https://github.com/atomist/npm-automations/blob/master/src/commands/simple/HelloWorld.ts (HelloWorld Command Handler)

Feel free to modify the code in the `HelloWorld` command handler,
restart your local automation client, and see what happens!

## Triggering an event handler

While command handlers respond to commands you send the Atomist bot,
_event handlers_ take action when different types of events occur in
your development and operations environment.  Some examples of events
are commits are pushed to a repo, or a CI build fails, or an instance
of a running service becomes unhealthy.  Example responses to those
events are showing the commits in a Slack message, automatically
restarting the build, and triggering a PagerDuty alert, respectively.

The sample event handler in this project will notice when a new issue
is created in the [atomist-rugs][] GitHub organization and then send a
Slack message and post its own comment on the issue.  You can trigger
the [`CommentOnIssue`][issue-handler] event handler by
simply
[creating an issue in the atomist-rugs/cd-test-01 repo][create-issue].
You should see that event appear in the console logs of your locally
running automation client, followed by a log of the actions the event
handler is taking.  Once those actions are complete, you should see a
new message in the `#cd-test-01` channel in the testing Slack team and
a new comment on the issue you just created.

[issue-handler]: https://github.com/atomist/npm-automations/blob/master/src/events/CommentOnIssue.ts (CommentOnIssue Event Handler)
[create-issue]: https://github.com/atomist-rugs/cd-test-01/issues/new (Create Issue in atomist-rugs/cd-test-01)

## Dashboard and GraphQL data explorer

When the automation client has successfully established a connection
to the Atomist API server, the Dashboard (work-in-progress) and
GraphQL data explorer will be available.

*   Dashboard: http://localhost:2866
*   GraphQL Data Explorer: http://localhost:2866/graphql

## Support

General support questions should be discussed in the `#support`
channel in our community Slack team
at [atomist-community.slack.com][slack].

If you find a problem, please create an [issue][].

[issue]: https://github.com/atomist/npm-automations/issues

## Development

You will need to install [node][] to build and test this project.

### Build and Test

Command | Reason
------- | ------
`npm install` | to install all the required packages
`npm run start` | to start the Atomist automation client
`npm run lint` | to run tslint against the TypeScript
`npm run compile` | to compile all TypeScript into JavaScript
`npm test` | to run tests and ensure everything is working
`npm run autotest` | run tests continuously (you may also need to run `tsc -w`)
`npm run clean` | remove stray compiled JavaScript files and build directory

### Release

To create a new release of the project, simply push a tag of the form
`M.N.P` where `M`, `N`, and `P` are integers that form the next
appropriate [semantic version][semver] for release.  The version in
the package.json is replaced by the build and is totally ignored!  For
example:

[semver]: http://semver.org

```
$ git tag -a 1.2.3
$ git push --tags
```

The Travis CI build (see badge at the top of this page) will publish
the NPM module and automatically create a GitHub release using the tag
name for the release and the comment provided on the annotated tag as
the contents of the release notes.

---

Created by [Atomist][atomist].
Need Help?  [Join our Slack team][slack].

[atomist]: https://www.atomist.com/
[slack]: https://join.atomist.com


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][]

[Unreleased]: https://github.com/atomist/npm-automations/compare/0.1.0...HEAD

## [0.1.0][] - 2017-10-08

Initial Release

[0.1.0]: https://github.com/atomist/npm-automations/tree/0.1.0

### Added

-   Add commit status with prerelease module URL


test/util/tempProject.ts
/*
import { LocalProject } from "@atomist/automation-client/project/local/LocalProject";
import { NodeFsLocalProject } from "@atomist/automation-client/project/local/NodeFsLocalProject";
import * as tmp from "tmp";

export function tempProject(): LocalProject {
    const dir = tmp.dirSync();
    return new NodeFsLocalProject("temp", dir.name);
}
*/


test/events/BuildPushStatusTest.ts
/*
 * Copyright © 2017 Atomist, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import "mocha";
import * as assert from "power-assert";

import { EventFired } from "@atomist/automation-client/Handlers";

import { findBuildTags, postStatusToGitHub } from "../../src/events/BuildPublishStatus";
import { BuildWithRepo } from "../../src/typings/types";

import { AxiosPromise, AxiosRequestConfig, AxiosResponse, AxiosStatic } from "axios";

describe("findBuildTags", () => {

    const buildPlatform = "travis";

    it("should find a push build tag and return the version portion of it", () => {
        const v = "23.29.31";
        const prv = `${v}-37414347535961`;
        const buildName = "67";
        const tags: BuildWithRepo.Tags[] = [
            { name: `${prv}+${buildPlatform}.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "push", v, buildName, buildPlatform);
        assert(buildTags.length === 1);
        assert(buildTags[0] === prv);
    });

    it("should find a push build tag among a list", () => {
        const v = "23.29.31";
        const prv = `${v}-37414347535961`;
        const buildName = "67";
        const tags: BuildWithRepo.Tags[] = [
            { name: `1${prv}+${buildPlatform}.${buildName}` },
            { name: `${prv}+${buildPlatform}.${buildName}` },
            { name: `${prv}+${buildPlatform}.${buildName}1` },
            { name: `${prv}+circleci.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "push", v, buildName, buildPlatform);
        assert(buildTags.length === 1);
        assert(buildTags[0] === prv);
    });

    it("should find a push build tags among a list", () => {
        const v = "23.29.31";
        const tss = ["37414347535961", "41434753596171", "43475359617173"];
        const prvs = tss.map(t => `${v}-${t}`);
        const buildName = "79";
        const tags: BuildWithRepo.Tags[] = [
            { name: `1${prvs[0]}+${buildPlatform}.${buildName}` },
            { name: `${prvs[0]}+${buildPlatform}.${buildName}` },
            { name: `${prvs[1]}+${buildPlatform}.${buildName}1` },
            { name: `${prvs[1]}+${buildPlatform}.${buildName}` },
            { name: `${prvs[2]}+${buildPlatform}.${buildName}` },
            { name: `${prvs[2]}+circleci.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "push", v, buildName, buildPlatform);
        assert(buildTags.length === prvs.length);
        prvs.forEach((bv, i) => assert(buildTags[i] === bv));
    });

    it("should not find a push build tag if one does not exist", () => {
        const v = "23.29.31";
        const prv = `${v}-37414347535961`;
        const buildName = "67";
        const tags: BuildWithRepo.Tags[] = [
            { name: `1${prv}+${buildPlatform}.${buildName}` },
            { name: `${prv}+${buildPlatform}.${buildName}1` },
            { name: `${prv}+circleci.${buildName}1` },
            { name: `${prv}-branch+${buildPlatform}.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "push", v, buildName, buildPlatform);
        assert(buildTags.length === 0);
    });

    it("should find a PR build tag and return the version portion of it", () => {
        const v = "23.29.31";
        const branch = "prime-movers";
        const prv = `${v}-${branch}.37414347535961`;
        const buildName = "67";
        const tags: BuildWithRepo.Tags[] = [
            { name: `${prv}+${buildPlatform}.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "pull_request", v, buildName, buildPlatform, branch);
        assert(buildTags.length === 1);
        assert(buildTags[0] === prv);
    });

    it("should find a PR build tag using safe branch name", () => {
        const v = "23.29.31";
        const badBranch = "pr1me_.mov3rs-+1";
        const branch = "pr1me.mov3rs-.1";
        const prv = `${v}-${branch}.37414347535961`;
        const buildName = "67";
        const tags: BuildWithRepo.Tags[] = [
            { name: `${prv}+${buildPlatform}.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "pull_request", v, buildName, buildPlatform, badBranch);
        assert(buildTags.length === 1);
        assert(buildTags[0] === prv);
    });

    it("should find a PR build tag among a list", () => {
        const v = "23.29.31";
        const branch = "prime-movers";
        const prv = `${v}-${branch}.37414347535961`;
        const buildName = "67";
        const tags: BuildWithRepo.Tags[] = [
            { name: `1${prv}+${buildPlatform}.${buildName}` },
            { name: `${prv}+${buildPlatform}.${buildName}` },
            { name: `${prv}+${buildPlatform}.${buildName}1` },
            { name: `${prv}+circleci.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "pull_request", v, buildName, buildPlatform, branch);
        assert(buildTags.length === 1);
        assert(buildTags[0] === prv);
    });

    it("should find a PR build tags among a list", () => {
        const v = "23.29.31";
        const branch = "prime-movers";
        const tss = ["37414347535961", "41434753596171", "43475359617173"];
        const prvs = tss.map(t => `${v}-${branch}.${t}`);
        const buildName = "79";
        const tags: BuildWithRepo.Tags[] = [
            { name: `1${prvs[0]}+${buildPlatform}.${buildName}` },
            { name: `${prvs[0]}+${buildPlatform}.${buildName}` },
            { name: `${prvs[1]}+${buildPlatform}.${buildName}1` },
            { name: `${prvs[1]}+${buildPlatform}.${buildName}` },
            { name: `${prvs[2]}+${buildPlatform}.${buildName}` },
            { name: `${prvs[2]}+circleci.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "pull_request", v, buildName, buildPlatform, branch);
        assert(buildTags.length === prvs.length);
        prvs.forEach((bv, i) => assert(buildTags[i] === bv));
    });

    it("should find not a PR build tag if no branch given", () => {
        const v = "23.29.31";
        const branch = "prime-movers";
        const prv = `${v}-${branch}.37414347535961`;
        const buildName = "67";
        const tags: BuildWithRepo.Tags[] = [
            { name: `${prv}+${buildPlatform}.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "pull_request", v, buildName, buildPlatform);
        assert(buildTags.length === 0);
    });

    it("should not find a PR build tag if one does not exist", () => {
        const v = "23.29.31";
        const branch = "prime-movers";
        const prv = `${v}-${branch}.37414347535961`;
        const buildName = "67";
        const tags: BuildWithRepo.Tags[] = [
            { name: `1${prv}+${buildPlatform}.${buildName}` },
            { name: `${prv}+${buildPlatform}.${buildName}1` },
            { name: `${prv}+circleci.${buildName}1` },
            { name: `${prv}-branch+${buildPlatform}.${buildName}` },
        ];
        const buildTags = findBuildTags(tags, "pull_request", v, buildName, buildPlatform, branch);
        assert(buildTags.length === 0);
    });

});

describe("postStatusToGitHub", () => {

    /* tslint:disable:max-line-length */
    const getPkgJsonResponse = {
        name: "package.json",
        path: "package.json",
        sha: "8a0a4a41f5297e10290045fd102a01013dd82040",
        size: 1621,
        url: "https://api.github.com/repos/atomist/slack-messages/contents/package.json?ref=ac5de03114a51fba54f39143be49ea37e075faec",
        html_url: "https://github.com/atomist/slack-messages/blob/ac5de03114a51fba54f39143be49ea37e075faec/package.json",
        git_url: "https://api.github.com/repos/atomist/slack-messages/git/blobs/8a0a4a41f5297e10290045fd102a01013dd82040",
        download_url: "https://raw.githubusercontent.com/atomist/slack-messages/ac5de03114a51fba54f39143be49ea37e075faec/package.json",
        type: "file",
        content: "ewogICJuYW1lIjogIkBhdG9taXN0L3NsYWNrLW1lc3NhZ2VzIiwKICAiZGVz\nY3JpcHRpb24iOiAidXRpbGl0aWVzIGZvciBjcmVhdGluZyBTbGFjayBtZXNz\nYWdlcyIsCiAgInZlcnNpb24iOiAiMC4xMi4wIiwKICAiYXV0aG9yIjogIkF0\nb21pc3QiLAogICJsaWNlbnNlIjogIkFwYWNoZS0yLjAiLAogICJob21lcGFn\nZSI6ICJodHRwczovL2dpdGh1Yi5jb20vYXRvbWlzdC9zbGFjay1tZXNzYWdl\ncyNyZWFkbWUiLAogICJyZXBvc2l0b3J5IjogewogICAgInR5cGUiOiAiZ2l0\nIiwKICAgICJ1cmwiOiAiZ2l0K2h0dHBzOi8vZ2l0aHViLmNvbS9hdG9taXN0\nL3NsYWNrLW1lc3NhZ2VzLmdpdCIKICB9LAogICJrZXl3b3JkcyI6IFsKICAg\nICJhdG9taXN0IiwKICAgICJydWciLAogICAgInNsYWNrIgogIF0sCiAgImJ1\nZ3MiOiB7CiAgICAidXJsIjogImh0dHBzOi8vZ2l0aHViLmNvbS9hdG9taXN0\nL3NsYWNrLW1lc3NhZ2VzL2lzc3VlcyIKICB9LAogICJkZXBlbmRlbmNpZXMi\nOiB7CiAgICAiZGVwcmVjYXRlZC1kZWNvcmF0b3IiOiAiXjAuMS42IiwKICAg\nICJsb2Rhc2giOiAiXjQuMTcuNCIKICB9LAogICJkZXZEZXBlbmRlbmNpZXMi\nOiB7CiAgICAiQHR5cGVzL2xvZGFzaCI6ICJeNC4xNC43MyIsCiAgICAiQHR5\ncGVzL21vY2hhIjogIl4yLjIuNDEiLAogICAgIkB0eXBlcy9wb3dlci1hc3Nl\ncnQiOiAiXjEuNC4yOSIsCiAgICAiZXNwb3dlci10eXBlc2NyaXB0IjogIl44\nLjAuMCIsCiAgICAibW9jaGEiOiAiXjMuNC4yIiwKICAgICJwb3dlci1hc3Nl\ncnQiOiAiXjEuNC40IiwKICAgICJzdXBlcnZpc29yIjogIl4wLjEyLjAiLAog\nICAgInRzbGludCI6ICJeNS42LjAiLAogICAgInR5cGVkb2MiOiAiXjAuOC4w\nIiwKICAgICJ0eXBlc2NyaXB0IjogIl4yLjUuMSIsCiAgICAidHlwZXNjcmlw\ndC1mb3JtYXR0ZXIiOiAiXjYuMC4wIgogIH0sCiAgImRpcmVjdG9yaWVzIjog\newogICAgInRlc3QiOiAidGVzdCIKICB9LAogICJzY3JpcHRzIjogewogICAg\nImF1dG90ZXN0IjogInN1cGVydmlzb3IgLXEgLW4gZXhpdCAteCBucG0gLS0g\ndGVzdCIsCiAgICAiYnVpbGQiOiAibnBtIHJ1biBsaW50ICYmIG5wbSBydW4g\nY29tcGlsZSAmJiBucG0gdGVzdCIsCiAgICAiY2xlYW4iOiAibnBtIHJ1biBj\nbGVhbi1qcyA7IHJtIC1yZiBidWlsZCIsCiAgICAiY2xlYW4tanMiOiAiZmlu\nZCBzcmMgdGVzdCAtdHlwZSBmIC1uYW1lICcqLmpzJyAtcHJpbnQwIHwgeGFy\nZ3MgLTAgcm0gLWYiLAogICAgImNvbXBpbGUiOiAidHNjIC1wIC4iLAogICAg\nImRpc3RjbGVhbiI6ICJucG0gcnVuIGNsZWFuIDsgcm0gLXJmIG5vZGVfbW9k\ndWxlcyIsCiAgICAiZm10IjogInRzZm10IC0tcmVwbGFjZSIsCiAgICAibGlu\ndCI6ICJ0c2xpbnQgLS1mb3JtYXQgdmVyYm9zZSAtLXByb2plY3QgLiAtLWV4\nY2x1ZGUgJ3tidWlsZCxub2RlX21vZHVsZXN9LyoqJyAnKiovKi50cyciLAog\nICAgImxpbnQtZml4IjogIm5wbSBydW4gbGludCAtLSAtLWZpeCIsCiAgICAi\ndGVzdCI6ICJtb2NoYSAtLWNvbXBpbGVycyB0czplc3Bvd2VyLXR5cGVzY3Jp\ncHQvZ3Vlc3MgJ3Rlc3QvKiovKi50cyciLAogICAgInR5cGVkb2MiOiAidHlw\nZWRvYyAtLW1vZGUgbW9kdWxlcyAtLWV4Y2x1ZGVFeHRlcm5hbHMiCiAgfQp9\nCg==\n",
        encoding: "base64",
        _links: {
            self: "https://api.github.com/repos/atomist/slack-messages/contents/package.json?ref=ac5de03114a51fba54f39143be49ea37e075faec",
            git: "https://api.github.com/repos/atomist/slack-messages/git/blobs/8a0a4a41f5297e10290045fd102a01013dd82040",
            html: "https://github.com/atomist/slack-messages/blob/ac5de03114a51fba54f39143be49ea37e075faec/package.json",
        },
    };
    const pkgJson = {
        name: "@atomist/slack-messages",
        description: "utilities for creating Slack messages",
        version: "0.12.0",
        author: "Atomist",
        license: "Apache-2.0",
        homepage: "https://github.com/atomist/slack-messages#readme",
        repository: {
            type: "git",
            url: "git+https://github.com/atomist/slack-messages.git",
        },
        keywords: [
            "atomist",
            "rug",
            "slack",
        ],
        bugs: {
            url: "https://github.com/atomist/slack-messages/issues",
        },
        dependencies: {
            "deprecated-decorator": "^0.1.6",
            "lodash": "^4.17.4",
        },
        devDependencies: {
            "@types/lodash": "^4.14.73",
            "@types/mocha": "^2.2.41",
            "@types/power-assert": "^1.4.29",
            "espower-typescript": "^8.0.0",
            "mocha": "^3.4.2",
            "power-assert": "^1.4.4",
            "supervisor": "^0.12.0",
            "tslint": "^5.6.0",
            "typedoc": "^0.8.0",
            "typescript": "^2.5.1",
            "typescript-formatter": "^6.0.0",
        },
        directories: {
            test: "test",
        },
        scripts: {
            "autotest": "supervisor -q -n exit -x npm -- test",
            "build": "npm run lint && npm run compile && npm test",
            "clean": "npm run clean-js ; rm -rf build",
            "clean-js": "find src test -type f -name '*.js' -print0 | xargs -0 rm -f",
            "compile": "tsc -p .",
            "distclean": "npm run clean ; rm -rf node_modules",
            "fmt": "tsfmt --replace",
            "lint": "tslint --format verbose --project . --exclude '{build,node_modules}/**' '**/*.ts'",
            "lint-fix": "npm run lint -- --fix",
            "test": "mocha --compilers ts:espower-typescript/guess 'test/**/*.ts'",
            "typedoc": "typedoc --mode modules --excludeExternals",
        },
    };

    const buildEvent = {
        data: {
            Build: [
                {
                    push: {
                        branch: "master",
                    },
                    pullRequestNumber: undefined,
                    name: "1190",
                    pullRequest: undefined,
                    status: "passed",
                    commit: {
                        sha: "ad31a1182a194c09b960d75b0f4002be1bbca288",
                        statuses: [
                            {
                                context: "fingerprint/atomist",
                                description: "No blocking Fingerprint changes",
                                targetUrl: "",
                            },
                            {
                                context: "continuous-integration/travis-ci/push",
                                description: "The Travis CI build failed",
                                targetUrl: "https://travis-ci.org/atomist/slack-messages/builds/280762398?utm_source=github_status&utm_medium=notification",
                            },
                        ],
                        tags: [
                            {
                                name: "0.12.0-20171006123456+travis.1190",
                            },
                        ],
                        author: {
                            login: "cdupuis",
                            person: {
                                chatId: {
                                    screenName: "cd",
                                    userId: "U1L22E3SA",
                                },
                            },
                        },
                    },
                    repo: {
                        owner: "atomist",
                        name: "slack-messages",
                        defaultBranch: "master",
                        channels: [
                            {
                                name: "slack-messages",
                            },
                        ],
                        org: {
                            provider: undefined,
                        },
                    },
                    buildUrl: "https://travis-ci.org/atomist/slack-messages/builds/280762398",
                    buildId: "280762398",
                    trigger: "push",
                    provider: "travis",
                },
            ],
        },
        extensions: {
            type: "READ_ONLY",
            operationName: "BuildWithRepo",
        },
    } as EventFired<BuildWithRepo.Subscription>;

    function fakeAxios(getStatus: number = 200, postStatus: number = 200): AxiosStatic {
        return {
            get(url: string, config?: AxiosRequestConfig): AxiosPromise {
                const ar = {
                    status: getStatus,
                    data: getPkgJsonResponse,
                    statusText: "Status Text",
                    headers: { header: "Header" },
                } as AxiosResponse;
                return Promise.resolve(ar);
            },
            post(url: string, data?: any, config?: AxiosRequestConfig): AxiosPromise {
                const ar = {
                    status: postStatus,
                    data: getPkgJsonResponse,
                    statusText: "Status Text",
                    headers: { header: "Header" },
                } as AxiosResponse;
                return Promise.resolve(ar);
            },
        } as AxiosStatic;
    }

    it("should be successful", done => {
        const promise = postStatusToGitHub(buildEvent, "dummy-token", fakeAxios());
        promise.then(result => {
            assert(result.code === 0);
            assert((result as any).message === "posted status to GitHub");
            done();
        }).catch(done);
    });

    it("should fail when getting package.json errors", done => {
        const promise = postStatusToGitHub(buildEvent, "dummy-token", fakeAxios(500));
        promise.then(result => {
            assert(result.code === 1);
            assert((result as any).error.indexOf("get atomist/slack-messages/package.json failed") > 0);
            done();
        }).catch(done);
    });

    it("should succeed when there is no package.json", done => {
        const promise = postStatusToGitHub(buildEvent, "dummy-token", fakeAxios(404));
        promise.then(result => {
            assert(result.code === 0);
            assert((result as any).message === "not an NPM package repository");
            done();
        }).catch(done);
    });

    it("should fail when posting the status fails", done => {
        const promise = postStatusToGitHub(buildEvent, "dummy-token", fakeAxios(200, 500));
        promise.then(result => {
            assert(result.code === 1);
            assert((result as any).error.indexOf("failed to post to https://api.github.com/repos") > 0);
            done();
        }).catch(done);
    });

});


src/events/BuildPublishStatus.ts
/*
 * Copyright © 2017 Atomist, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { subscriptionFromFile } from "@atomist/automation-client/graph/graphQL";
import { Failure, Success } from "@atomist/automation-client/HandlerResult";
import {
    EventFired,
    EventHandler,
    HandleEvent,
    HandlerContext,
    HandlerResult,
    Secret,
    Secrets,
    Tags,
} from "@atomist/automation-client/Handlers";
import { logger } from "@atomist/automation-client/internal/util/logger";

import axios, { AxiosResponse, AxiosStatic } from "axios";

import { BuildWithRepo } from "../typings/types";

interface CommitStatusPayload {
    state: "error" | "failure" | "pending" | "success";
    target_url?: string;
    description?: string;
    context?: string;
}

interface ErrorHandlerResult extends HandlerResult {
    code: number;
    error: string;
}

function errorResult(msg: string): HandlerResult {
    const e: ErrorHandlerResult = { code: 1, error: msg };
    return e as HandlerResult;
}

interface SuccessHandlerResult extends HandlerResult {
    code: 0;
    message: string;
}

function successResult(msg: string): HandlerResult {
    const s: SuccessHandlerResult = { code: 0, message: msg };
    return s as HandlerResult;
}

@EventHandler("create commit status with NPM module URL after successful publication",
    subscriptionFromFile("graphql/build"))
@Tags("build", "npm")
export class BuildPublishStatus implements HandleEvent<BuildWithRepo.Subscription> {

    @Secret(Secrets.ORG_TOKEN)
    public githubToken: string;

    public handle(e: EventFired<BuildWithRepo.Subscription>, ctx: HandlerContext): Promise<HandlerResult> {
        logger.info(`incoming event is %s`, JSON.stringify(e.data, null, 2));

        return postStatusToGitHub(e, this.githubToken, axios);
    }

}

/**
 * Make separate function so we can mock and test.
 */
export function postStatusToGitHub(
    e: EventFired<BuildWithRepo.Subscription>,
    token: string,
    axon: AxiosStatic,
): Promise<HandlerResult> {

    if (e && e.data && e.data.Build) {
        const axiosConfig = {
            headers: {
                "Accept": "application/vnd.github.v3+json",
                "Authorization": `token ${token}`,
                "Content-Type": "application/json",
            },
        };

        return Promise.all(e.data.Build.map(build => {
            if (build.status !== "passed") {
                return successResult(`build did not pass`);
            }
            if (!build.name || !build.repo || !build.trigger || !build.commit ||
                !build.commit.sha || !build.commit.tags) {
                return successResult(`build event does not have needed properties to post status`);
            }

            const buildName = build.name;
            const repo = build.repo;
            const slug = `${repo.owner}/${repo.name}`;
            const trigger = build.trigger;
            const commit = build.commit;
            const sha = build.commit.sha;
            const tags = build.commit.tags;
            const defaultBranch = (repo.defaultBranch) ? repo.defaultBranch : "master";
            const isPushToDefaultBranchBuild = build.trigger === "push" &&
                build.push && build.push.branch === defaultBranch;
            const isPullRequestBuild = build.trigger === "pull_request"
                && build.pullRequest && build.pullRequest.branchName;
            if (!isPushToDefaultBranchBuild && !isPullRequestBuild) {
                const msg = `build ${slug}#${buildName} is neither on default branch nor a PR`;
                logger.info(msg);
                return successResult(msg);
            }

            const ghApiBase = (repo.org && repo.org.provider && repo.org.provider.apiUrl) ?
                repo.org.provider.apiUrl : "https://api.github.com/";
            const ghPkgJsonUrl = `${ghApiBase}repos/${slug}/contents/package.json?ref=${build.commit.sha}`;
            return axon.get(ghPkgJsonUrl, axiosConfig)
                .then(pkgJsonResponse => {
                    console.log(``);
                    if (pkgJsonResponse.status === 404) {
                        logger.info(`failed to find ${slug}/package.json: ${httpErrorMsg(pkgJsonResponse)}`);
                        return successResult(`not an NPM package repository`);
                    } else if (pkgJsonResponse.status < 200 || pkgJsonResponse.status > 299) {
                        const msg = `get ${slug}/package.json failed: ${httpErrorMsg(pkgJsonResponse)}`;
                        logger.error(msg);
                        return errorResult(msg);
                    }

                    const pkgJsonContentBase64 = pkgJsonResponse.data.content;
                    const pkgJsonBuffer = new Buffer(pkgJsonContentBase64, "base64");
                    const pkgJsonContent = pkgJsonBuffer.toString();
                    const pkgJson = JSON.parse(pkgJsonContent);
                    const moduleName = pkgJson.name;
                    const moduleBaseVersion = pkgJson.version;

                    const branchName = (build.pullRequest && build.pullRequest.branchName) ?
                        build.pullRequest.branchName : undefined;
                    const ghStatusUrl = `${ghApiBase}repos/${slug}/statuses/${sha}`;
                    const artifactoryBase = "https://atomist.jfrog.io/atomist/npm-dev";
                    const moduleBase = `${artifactoryBase}/${moduleName}/-/${moduleName}`;

                    const moduleVersions = findBuildTags(tags, trigger, moduleBaseVersion,
                        buildName, "travis", branchName);

                    return Promise.all(moduleVersions.map(moduleVersion => {

                        const moduleUrl = `${moduleBase}-${moduleVersion}.tgz`;
                        const statusData: CommitStatusPayload = {
                            state: "success",
                            description: "Prerelease NPM module publication",
                            context: "npm/module/prerelease",
                            target_url: moduleUrl,
                        };

                        return axon.post(ghStatusUrl, statusData, axiosConfig)
                            .then(statusResponse => {
                                if (statusResponse.status < 200 || statusResponse.status > 299) {
                                    const msg = `failed to post to ${ghStatusUrl}: ${httpErrorMsg(statusResponse)}`;
                                    logger.error(msg);
                                    return errorResult(msg);
                                }
                                return successResult(`posted status to GitHub`);
                            })
                            .catch((err: any) =>
                                errorResult(`failed to post to ${ghStatusUrl}: ${(err as Error).message}`));
                    }))
                        .then(resolved => resolveManyToOne(resolved))
                        .catch((err: any) => errorResult((err as Error).message));
                });
        }))
            .then(resolved => resolveManyToOne(resolved))
            .catch((err: any) => errorResult((err as Error).message));
    }

    const noOpMsg = `BuildPublishStatus: build event fired but event had no build`;
    logger.warn(noOpMsg);
    return Promise.resolve(errorResult(noOpMsg));
}

/**
 * If any of the results are Failure, return a Failure with a
 * concatenation of all available error messages.
 *
 * @param results array of handler results, typically returned by Promise.all
 * @return Success if all results are Success, failure otherwise
 */
function resolveManyToOne(results: HandlerResult[]): HandlerResult {
    const failures = results.filter(r => r.code !== 0);
    if (failures.length > 0) {
        const errors = failures.filter(f => (f as any).error) as ErrorHandlerResult[];
        const msg = `processing of some events failed: ${errors.map(e => e.error).join("; ")}`;
        return errorResult(msg);
    }
    return (results.length === 1) ? results[0] : successResult(`all promises passed`);
}

function httpErrorMsg(r: AxiosResponse): string {
    return `${r.statusText}: ${JSON.stringify(r.headers)}`;
}

/**
 * Find a timestamped prerelease version tag from list of tags.
 *
 * @param tags array of tags
 * @param trigger what triggered the build, "push" or "pull_request"
 * @param version base version of package
 * @param buildName name of build
 * @param buildPlatform CI provider, e.g., "travis"
 * @param branch branch of build, only used for PR builds
 * @return array of tags, may be empty
 */
export function findBuildTags(
    tags: BuildWithRepo.Tags[],
    trigger: "push" | "pull_request",
    version: string,
    buildName: string,
    buildPlatform: string = "travis",
    branch?: string,
): string[] {

    const buildMeta = `${buildPlatform}\\.${buildName}`;
    let tagRegExp: RegExp;
    if (trigger === "push") {
        tagRegExp = new RegExp(`^${version}-\\d{14}\\+${buildMeta}\$`);
    } else if (branch) {
        const safeBranch = branch.replace(/[^A-Za-z0-9\\-]+/g, ".");
        tagRegExp = new RegExp(`^${version}-${safeBranch}\\.\\d{14}\\+${buildMeta}\$`);
    } else {
        return [];
    }

    return (tags.map(t => t.name).filter(t => !!t && tagRegExp.test(t)) as string[])
        .map(t => t && t.replace(/\+.*$/, ""));
}
